### ISSUE TITLE
Spans returned by `ast` are incorrect

### ISSUE THREAD

**Issue by NotTheDr01ds** [nushell/nushell#12945]
### Describe the bug

Not sure if this is a bug or enhancement, honestly.  Since `ast` returns span information, I think there's an assumption that it actually exists, so I'm siding with "bug" for the moment ;-)

When parsing with the `ast` command, each element in the tree is returned with span information, but the spans are not valid.

### How to reproduce

```nushell
> let span = ( ast 'true' --json | get block | from json | get span )
> view span $span.start $span.end
$spa
```

Although results will probably be indeterminate since the span isn't valid (may be part of another file or even outside the current engine state's filelist.

Also, of course, `view files` doesn't have an entry for the span in the first plan.

### Expected behavior

`ast` shouldn't return spans that aren't valid.  Either:

* `ast` shouldn't return spans at all
* (Preferably) `ast` should return valid spans for the parsed file

### Configuration

| key                | value                                           |
| ------------------ | ----------------------------------------------- |
| version            | 0.93.0                                          |
| major              | 0                                               |
| minor              | 93                                              |
| patch              | 0                                               |
| branch             |                                                 |
| commit_hash        |                                                 |
| build_os           | linux-x86_64                                    |
| build_target       | x86_64-unknown-linux-gnu                        |
| rust_version       | rustc 1.77.2 (25ef9e3d8 2024-04-09)             |
| rust_channel       | stable-x86_64-unknown-linux-gnu                 |
| cargo_version      | cargo 1.77.2 (e52e36006 2024-03-26)             |
| build_time         | 2024-05-04 12:29:25 -04:00                      |
| build_rust_channel | release                                         |
| allocator          | mimalloc                                        |
| features           | default, sqlite, system-clipboard, trash, which |
| installed_plugins  |                                                 |

### Additional context

Would it be possible to add a "fake file" like "Host Environment Variables" in `util.rs`?  Something like *"ast entry #1"*?

**Comment by NotTheDr01ds** [comment]
Related: #11687

**Comment by NotTheDr01ds** [comment]
> ast entry #1 is `source`

I *think* you mean that *"view files"* entry 1 (or 0) is `source`, right?

I probably should have been more clear.  And I really do hope I'm understanding the `view files` and `view spans` properly.  I *think* I'm finally grokking it after a long time of being clueless about how things worked.  Took some poking around in the source, crate doc, issues, and Discord messages to get my level-of-understanding to where it is ;-).

Right now, we have something like this in `view files`:

```nushell
╭────┬───────────────────────────────────────────────────────────────┬────────┬────────┬───────╮
│  # │                           filename                            │ start  │  end   │ size  │
├────┼───────────────────────────────────────────────────────────────┼────────┼────────┼───────┤
│  0 │ source                                                        │      0 │     10 │    10 │
│  1 │ Host Environment Variables                                    │     10 │   5055 │  5045 │
│  2 │ NU_STDLIB_VIRTUAL_DIR/std/mod.nu                              │   5055 │  12321 │  7266 │
...
│ 14 │ /home/ntd/.config/nushell/env.nu                              │  85651 │  85848 │   197 │
│ 15 │ /home/ntd/.config/nushell/env/00_init_libs.nu                 │  85848 │  86906 │  1058 │
│ 16 │ display_output hook                                           │  86906 │  86963 │    57 │
... # Many more config files
│ 45 │ show_banner                                                   │ 131439 │ 131461 │    22 │
│ 46 │ entry #1                                                      │ 131461 │ 131474 │    13 │
│ 47 │ entry #2                                                      │ 131474 │ 131522 │    48 │
│ 48 │ entry #4                                                      │ 131522 │ 131532 │    10 │
│ 49 │ entry #5                                                      │ 131532 │ 131542 │    10 │
├────┼───────────────────────────────────────────────────────────────┼────────┼────────┼───────┤
│  # │                           filename                            │ start  │  end   │ size  │
╰────┴───────────────────────────────────────────────────────────────┴────────┴────────┴───────╯
```

If I then:

```nushell
ast 'true'
```

... I then have `entry #6` which is the span for `ast 'true'` itself.

If there's a valid span for the ast *result`, then wouldn't it need to be in a "file" of its own?  E.g., `ast entry #1`?

**Comment by fdncred** [comment]
> I think you mean that "view files" entry 1 (or 0) is source, right?

yes. and if you look at this span, i think it's nu. mine is always 0 2. yours appears to be 0 10 which i assume is something like /usr/bin/nu.

`view files` are things loaded in memory. some are files, most will be command line entries in the repl. you can view any of them to see what they are. we also have the size here because this stuff is loaded in memory.

I'm not exactly sure why the spans aren't showing up right. My guess is because it's being parsed but it's not really part of the repl, so it's not recorded. Some digging in would need to take place.

**Comment by NotTheDr01ds** [comment]
> yours appears to be 0 10 which i assume is something like /usr/bin/nu.

Close, of course - `nu --login`.  I just wasn't sure why you brought it up, other than perhaps something I said originally led you to think I was asking about that.  I think it was my mentioning `ast entry #1`, which I just meant as a hypothetical naming that would follow the same convention as `entry #<n>`

> I'm not exactly sure why the spans aren't showing up right. My guess is because it's being parsed but it's not really part of the repl, so it's not recorded. Some digging in would need to take place.

Also pretty close, it seems.  From what I can tell, the parser itself [just updates a working set](https://github.com/nushell/nushell/blob/0b5a4c0d951622e60c318ecaa66f15ab2c80eeaa/crates/nu-parser/src/parser.rs#L6294), but it's actually the CLI's `evaluate_source()` that [merges that working set back into the EngineState](https://github.com/nushell/nushell/blob/bf07806b1bc96eebcaa57456c97035a6bce68ef7/crates/nu-cli/src/util.rs#L271).  Since the `ast` command obviously doesn't actually evaluate its argument, the tree isn't added to the EngineState.

And (since I was going to try a first-pass at a fix) since Nushell commands are only given an immutable reference to EngineState, `ast` can't modify the EngineState directly either.

Is that a close assessment?

**Comment by fdncred** [comment]
Ya, that's what I meant. :) Seems pretty close to me.

**Comment by NotTheDr01ds** [comment]
cc @kubouch as you are working on the new Span implementation.  As you mentioned in the Discord API-update today ...

> Spans are to be created only by the parser (and a very few other places) and stored into a vector in the EngineState.

This seems to make implementation of the `ast` command a bit difficult.

(Note: Not suggesting you fix `ast`; just that as you are looking at spans, any thoughts you have on `ast` would be welcome here)

**Comment by NotTheDr01ds** [comment]
@kubouch That's a fair point; if the new parser actually adds spans to the EngineState directly, then `ast` should work again.  Currently, the parser itself doesn't do this *directly* - It leaves the `WorkingState` merge to the CLI's [`evaluate_source()`](https://github.com/nushell/nushell/blob/bf07806b1bc96eebcaa57456c97035a6bce68ef7/crates/nu-cli/src/util.rs#L271), which misses the `ast` command since it's *argument* doesn't go through `evaluate_source()`, right?

Is the plan for the the new parser to rework this part? (I know that question may be a bit early, but it's why I'm trying to get this in front of you as you are working on it).

**Comment by kubouch** [comment]
The EngineState and StateWorkingSet will be changed, but we'll probably have someting similar to the current merging. We'll need to experiment a bit and see what works, it's too early to be more precise.

**Comment by Juhan280** [comment]
I don't know if its already known (probably is), but just for added context I get this output in version `0.107.0`
```nu
~
❯ ast -j "test" | get block | from json | get span
╭───────┬────────╮
│ start │ 180224 │
│ end   │ 180228 │
╰───────┴────────╯

~
❯ view span 180224 180228
view

~
❯ 'test" | get block | from json | get span' | str length
40

~
❯ view span (180224 - 40) (180228 - 40)
test

```
And this is what I get from the reproduction code mentioned in this issue
```nu
> let span = ( ast 'true' --json | get block | from json | get span )
> view span $span.start $span.end
view
```

Here it looks like the start index of the span is wrong. Precisely, the start index of the span is at the end of the span buffer. And everything else is offset from that.

**Comment by fdncred** [comment]
@Juhan280 I'd be interested to know how this problem with spans in the `ast` command compares to `--ide-ast` nushell command line parameter. Have you compared those? e.g. Does the same problem exist there too or is `--ide-ast` "right" here? Just thinking that it could lead somewhere if both results are different.

I'm going by memory here so it could be wrong but when we were writing the --ide-ast stuff I remember getting an offset and then adding that to something to get the "proper offset" for the spans. Maybe that's missing somewhere?

**Comment by Juhan280** [comment]
`nu --ide-ast` has similar output as `ast --flatten` with only difference being that `--ide-ast` has an extra `type` column whose value is always `ast`

![Screenshot_20251012-033650_Termux.png](https://github.com/user-attachments/assets/6e155311-378e-4587-8fba-a872b9ee1571)

**Comment by fdncred** [comment]
interesting that they both start at offset 0. i would've expected the one done in the repl to start at the next available offset.

it's interesting here that it appears the span is being set wrong in the `asj -j "test"` example.

**Comment by fdncred** [comment]
Seems like the pipeline span could be the right one.

<img width="1727" height="434" alt="Image" src="https://github.com/user-attachments/assets/c2dc45bc-f6ad-40db-a7ff-4830dd24bfce" />
